quantus/metrics/base_batched.py

```diff
@@ -0,0 +1,404 @@
+        >> a_batch_saliency = a_batch_saliency.cpu().numpy()
+
+        # Initialise the metric and evaluate explanations by calling the metric instance.
+        >> metric = Metric(abs=True, normalise=False)
+        >> scores = metric(model=model, x_batch=x_batch, y_batch=y_batch, a_batch=a_batch_saliency}
+        """
+        # Run deprecation warnings.
+        warn_func.deprecation_warnings(kwargs)
+        warn_func.check_kwargs(kwargs)
+
+        data = self.general_preprocess(
+            model=model,
+            x_batch=x_batch,
+            y_batch=y_batch,
+            a_batch=a_batch,
+            s_batch=s_batch,
+            custom_batch=custom_batch,
+            channel_first=channel_first,
+            explain_func=explain_func,
+            explain_func_kwargs=explain_func_kwargs,
+            model_predict_kwargs=model_predict_kwargs,
+            softmax=softmax,
+            device=device,
+        )
+
+        # create generator for generating batches
+        batch_generator = self.generate_batches(
+            data=data, batch_size=batch_size,
+            display_progressbar=self.display_progressbar,
+        )
+
+        # TODO: initialize correct length of last results
+        self.last_results = []
+        # We use a tailing underscore to prevent confusion with the passed parameters.
+        # TODO: rename '_batch'-suffix kwargs of __call__() method accordingly or else this will be confusing.
+        for data_batch in batch_generator:
+            result = self.evaluate_batch(**data_batch, **self.custom_evaluate_kwargs)
+            # TODO: put in correct list idx instead of extending
+            self.last_results.extend(result)
+
+        # Call post-processing
+        self.custom_postprocess(**data)
+
+        self.all_results.append(self.last_results)
+        return self.last_results
+
+    @abstractmethod
+    def evaluate_batch(
+            self,
+            model: ModelInterface,
+            x_batch: np.ndarray,
+            y_batch: np.ndarray,
+            a_batch: np.ndarray,
+            s_batch: Optional[np.ndarray] = None,
+    ):
+        raise NotImplementedError()
+
+    @staticmethod
+    def get_number_of_batches(n_instances: int, batch_size: int):
+        return math.ceil(n_instances / batch_size)
+
+    def generate_batches(
+            self,
+            data: Dict[str, Any],
+            batch_size: int,
+            display_progressbar: bool = False,
+    ):
+        n_instances = len(data['x_batch'])
+
+        single_value_kwargs = {}
+        batched_value_kwargs = {}
+        for key, value in list(data.items()):
+            # If data-value is not a Sequence or a string, create list of value with length of n_instances.
+            if not isinstance(value, (Sequence, np.ndarray)) or isinstance(value, str):
+                single_value_kwargs[key] = value
+
+            # Check if sequence has correct length.
```

Comment:
Do we need to fix this right now or can we separate this into another issue? I think it would be great if we could scope this outside.